### PR TITLE
[Playground] Helm+kubernetes+deployment

### PR DESCRIPTION
**Pull request by MakarkinSAkvelon**
New solution for Playground backend - with Kubernetes and Helm
New pipeline for Backend deployment: using HELM and Kubernetes

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/infrastructure/helm-backend/templates/go.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+apiVersion: apps/v1
```
Comment (olehborysevych) [inline playground/infrastructure/helm-backend/templates/go.yml:1]: please rename deployment files to something like "deployment-go"

File: .github/workflows/k8e-backend.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Setup GCP account
+        run: |
+             echo "${{ secrets.S3D_GCP_PLAYGROUND_SA_KEY }}" | base64 -d > /tmp/gcp_access.json
+             which gcloud
+             gcloud auth activate-service-account --project=${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }} --key-file=/tmp/gcp_access.json
+      - name: create static ip
+        run: gcloud compute addresses create ${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }}-static --global
+      - name: Update variables
+        run: |
+             echo "redis_ip: $(gcloud redis instances list --region us-central1 --format="value(host)"):6379" >> playground/infrastructure/helm-backend/values.yaml
```
[the remark this answers is not in the document]
Comment (olehborysevych) [reply .github/workflows/k8e-backend.yml:58]: TBD. Not needed for now

**Review by eantyshev — COMMENTED** [review]
Could we merge this with the existing Helm chart at playground/infrastructure/helm
and re-use some boilerplate?
Probably, it is reasonable to move Helm configurations into  playground/terraform, because we believe that eventually it is where all of our infrastructure belongs.
I would suggest that we use [subcharts](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/) to reuse existing Helm configurations

File: .github/workflows/build_playground_backend_kubernetes.yml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Update variables
+        run: |
+             echo "redis_ip: $(gcloud redis instances list --region us-central1 --format="value(host)"):6379" >> playground/infrastructure/helm-backend/values.yaml
+             echo "project_id: "${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}"" >> playground/infrastructure/helm-backend/values.yaml
+             echo "registry: "${{ secrets.PLAYGROUND_REGISTRY_NAME}}/${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}/playground-repository"" >> playground/infrastructure/helm-backend/values.yaml
+             echo "static_ip: $(gcloud compute addresses list --filter "NAME:${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}-static" --format="value(address)")" >> playground/infrastructure/helm-backend/values.yaml
+             echo "static_ip_name: ${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}-static" >> playground/infrastructure/helm-backend/values.yaml
+             cat playground/infrastructure/helm-backend/values.yaml
+      - name: Get K8s Config
+        run: gcloud container clusters get-credentials --region us-central1-a playground-examples
```
Comment (eantyshev) [inline .github/workflows/build_playground_backend_kubernetes.yml:65]: Another hard-coded variable: "playground-examples" could be also extracted from terraform output

File: .github/workflows/build_playground_backend_kubernetes.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        run: ./gradlew playground:backend:containers:scio:dockerTagPush -Pdocker-repository-root='${{ secrets.PLAYGROUND_REGISTRY_NAME}}/${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}/playground-repository' -Pdocker-tag=${{ env.DOCKERTAG }}
+      - name: Build And Push Router Backend
+        run: ./gradlew playground:backend:containers:router:dockerTagPush -Pdocker-repository-root='${{ secrets.PLAYGROUND_REGISTRY_NAME}}/${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}/playground-repository' -Pdocker-tag=${{ env.DOCKERTAG }}
+      - name: Database and index creation
+        run: |
+              gcloud app create --region=us-central
+              gcloud datastore databases create --region=us-central
+              gcloud app deploy playground/index.yaml --project=${{ secrets.GCP_PLAYGROUND_PROJECT_ID }}
+      - name: Backend setup
+        run: helm install backend playground/infrastructure/helm-backend
```
Comment (eantyshev) [inline .github/workflows/build_playground_backend_kubernetes.yml:85]: Do we need to verify if everything was brought up successfully via kubectl?

Comment (eantyshev) [reply .github/workflows/build_playground_backend_kubernetes.yml:85]: Because we don't want to hide unsuccessful k8s deployments behind "success" workflow result

**Comment by olehborysevych** [comment]
lgtm

**Review by damondouglas — CHANGES_REQUESTED** [review]
@pabloem / @MakarkinSAkvelon This was challenging to test on my end.  I had tried to rebuild playground in my own environment but there were too many hard-coded values specific to the existing environment that this took too much time.  Perhaps a README would help to guide someone how to install in a new Google Cloud project.

File: playground/terraform/README.md [hunk 5]
```diff
@@ -17,73 +17,89 @@
...
-## 0. Create GCS bucket for state
-
-```bash
-$ gsutil mb -p ${PROJECT_ID} gs://state-bucket-name
-$ gsutil versioning set on gs://state-bucket-name
+# Prerequisit:
+
+* GCP project should be created
+* Bucket should be created
+* Service account with the following roles should be created:
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:24]: No we don't

File: playground/terraform/README.md [hunk 6]
```diff
@@ -17,73 +17,89 @@
...
+```
+# Infrastructure deployment:
+* Run following command for infrastructure deployment (please be sure that you are in the "beam" folder):
+```
+./gradlew playground:terraform:InitInfrastructure -Pproject_environment="beta"
+```
+# Backend deployment:
+* Login to Docker registry:
+```
+ cat `your json key locaton` | docker login -u _json_key --password-stdin https://`chosen_region`-docker.pkg.dev
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:86]: https://cloud.google.com/container-registry/docs/advanced-authentication (authentication using JSON file)

File: playground/terraform/README.md [hunk 7]
```diff
@@ -17,73 +17,89 @@
...
+    export GOOGLE_APPLICATION_CREDENTIALS=`your json key locaton`
 ```
-
-* Вeploy backend services
-
-```bash
-./gradlew playground:terraform:deployBackend -Pproject_environment="env-name" -Pdocker-tag="tag"
+* Activate created service account using following command:
+```
+    gcloud auth activate-service-account `full principal service account` --key-file=`your json key locaton`
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:64]: https://cloud.google.com/container-registry/docs/advanced-authentication

File: playground/terraform/README.md [hunk 8]
```diff
@@ -17,73 +17,89 @@
...
-
-## 3. Deploy application
-
-To deploy application run following steps:
-
-* Authinticate in Artifact registry
-
-```bash
-gcloud auth configure-docker us-central1-docker.pkg.dev
+    export GOOGLE_APPLICATION_CREDENTIALS=`your json key locaton`
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:60]: path could be "any" - absolute or relative
Link to "How create an account"?

File: playground/terraform/environment/beta/state.tfbackend [hunk 9]
```diff
@@ -17,4 +17,4 @@
 # under the License.
 #
 
-bucket  = "beam_playground_terraform"
+bucket  = "pg-second"
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/environment/beta/state.tfbackend:20]: any name, doesn't matter

File: playground/infrastructure/helm-backend/values.yaml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+replicaCount: 1
+image:
+   java_image: beam_playground-backend-java
+   go_image: beam_playground-backend-go
+   router_image: beam_playground-backend-router
+   scio_image: beam_playground-backend-scio
+   python_image: beam_playground-backend-python
+   pullPolicy: Always
+   tag: "beta"
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/infrastructure/helm-backend/values.yaml]: it's a variables in HELM, I don't think that we need to describe hos it's work, because they shouldn't be changed by hands

File: playground/infrastructure/helm-backend/values.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+service:
+   type: NodePort
+   targetPort: 8080
+   port: 443
+
+autoscaling:
+   enabled: false
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/infrastructure/helm-backend/values.yaml:32]: we're not ready yet (with scaling), moreover, we have only one GKE node for now

File: .github/workflows/playground_deploy_backend_gke.yml [hunk 12]
```diff
@@ -0,0 +1,65 @@
+name: kuber-bkend
+on:
+  push:
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply .github/workflows/playground_deploy_backend_gke.yml]: Completed

File: playground/terraform/README.md [hunk 13]
```diff
@@ -17,73 +17,89 @@
...
+```
+gcloud container clusters get-credentials --region `chosen_pg_location` `gke_name` --project `project_id`
+```
+* Database index creation:
+```
+gcloud app deploy playground/index.yaml --project=`project_id`
+```
+* Run following command for backend deployment (please be sure that you are in the "beam" folder):
+```
+./gradlew playground:terraform:gkebackend -Pproject_environment="beta" -Pdocker-tag="beta"
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Yes. we can use different tag for images (when we will start to create images for Example check)

File: playground/terraform/README.md [hunk 14]
```diff
@@ -16,74 +16,179 @@
...
-$ gsutil mb -p ${PROJECT_ID} gs://state-bucket-name
-$ gsutil versioning set on gs://state-bucket-name
+5. How to assign a new role for the service account:
+```
+https://cloud.google.com/iam/docs/granting-changing-revoking-access
 ```
+**Description:** * *It will explain how to assign required roles for your service account* *
 
-## 1. Create new environment
+6. How to install GCloud to your Operation system
```
Comment (AKosolapov) [inline playground/terraform/README.md:54]: I would rephrase: How to install gcloud CLI to your local environment

File: playground/terraform/README.md [hunk 15]
```diff
@@ -16,74 +16,179 @@
...
+**Description:** * *Describes the naming convention for Compute Engine resources.* *
+10. What is Docker
 ```
+https://docs.docker.com/get-started/overview/
+```
+**Description:** * *What is Docker and how to use it* *
 
-* Backend config
+***Google Cloud preparation steps:***
+When you create a new project, you must add a service account with a JSON key and a bucket according to the instructions above. Also, according to the instructions, configure the roles for the service account:
```
Comment (AKosolapov) [inline playground/terraform/README.md:84]: Consider rephrasing:
After you created or identified an existing project for deployment, add or select a service account with a JSON key, and add or select `bucket_name` for storing the Terraform state data according to the instructions above.

File: .github/workflows/playground_deploy_backend_gke.yml [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Build And Push Python Backend
+        run: ./gradlew playground:backend:containers:python:dockerTagPush -Pdocker-repository-root='${{ secrets.S3D_PLAYGROUND_REGISTRY_NAME}}/${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }}/playground-repository' -Pdocker-tag=${{ env.DOCKERTAG }}
+      - name: Build And Push Scio Backend
+        run: ./gradlew playground:backend:containers:scio:dockerTagPush -Pdocker-repository-root='${{ secrets.S3D_PLAYGROUND_REGISTRY_NAME}}/${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }}/playground-repository' -Pdocker-tag=${{ env.DOCKERTAG }}
+      - name: Build And Push Router Backend
+        run: ./gradlew playground:backend:containers:router:dockerTagPush -Pdocker-repository-root='${{ secrets.S3D_PLAYGROUND_REGISTRY_NAME}}/${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }}/playground-repository' -Pdocker-tag=${{ env.DOCKERTAG }}
+      - name: Index creation
+        run: |
+              gcloud app deploy playground/index.yaml --project=${{ secrets.S3D_GCP_PLAYGROUND_PROJECT_ID }}
```
Comment (olehborysevych) [inline .github/workflows/playground_deploy_backend_gke.yml:73]: @MakarkinSAkvelon this still uses your secret S3D

File: playground/terraform/build.gradle.kts [hunk 17]
```diff
@@ -395,3 +372,111 @@ task("deployBackend") {
...
+   var stdout = ByteArrayOutputStream()
+   if (project.hasProperty("docker-tag")) {
+        d_tag = project.property("docker-tag") as String
+   }
+   exec {
+       commandLine = listOf("terraform", "output", "playground_static_ip_address")
+       standardOutput = stdout
+   }
+   ipaddr = stdout.toString().trim().replace("\"", "")
+   stdout = ByteArrayOutputStream()
```
Comment (eantyshev) [inline playground/terraform/build.gradle.kts:393]: NIT this line belongs to `exec` statement below, not above, so let's group it more naturally

File: playground/terraform/build.gradle.kts [hunk 18]
```diff
@@ -395,3 +372,111 @@ task("deployBackend") {
...
+       commandLine = listOf("terraform", "output", "docker-repository-root")
+       standardOutput = stdout
+   }
+   registry = stdout.toString().trim().replace("\"", "")
+   stdout = ByteArrayOutputStream()
+   exec {
+       commandLine = listOf("terraform", "output", "playground_static_ip_address_name")
+       standardOutput = stdout
+   }
+   ipaddrname = stdout.toString().trim().replace("\"", "")
```
Comment (eantyshev) [inline playground/terraform/build.gradle.kts:417]: NIT Do we have functions to wrap this repetitive statements?

File: playground/terraform/build.gradle.kts [hunk 19]
```diff
@@ -395,3 +372,111 @@ task("deployBackend") {
     dependsOn(deploy)
 }
 
+task("takeConfig") {
```
Comment (eantyshev) [inline playground/terraform/build.gradle.kts:375]: NIT maybe, `setHelmValues`?

File: playground/terraform/build.gradle.kts [hunk 20]
```diff
@@ -395,3 +372,111 @@ task("deployBackend") {
...
+       standardOutput = stdout
+   }
+   registry = stdout.toString().trim().replace("\"", "")
+   stdout = ByteArrayOutputStream()
+   exec {
+       commandLine = listOf("terraform", "output", "playground_static_ip_address_name")
+       standardOutput = stdout
+   }
+   ipaddrname = stdout.toString().trim().replace("\"", "")
+   stdout = ByteArrayOutputStream()
```
Comment (eantyshev) [inline playground/terraform/build.gradle.kts:418]: NIT seems like this isn't needed

**Review by eantyshev — COMMENTED** [review]
@olehborysevych 
I suggest moving `helm-backend` to `playground/terraform`
`playground/infrastructure` is currently only about CI/CD,
and there'll be soon no k8s/helm

File: playground/terraform/infrastructure/variables.tf [hunk 21]
```diff
@@ -167,5 +165,8 @@ variable "gke_name" {
 
 variable "gke_location" {
   description = "Location of GKE cluster"
-  default     = "us-central1-a"
+}
+
+variable "pg_location" {
```
Comment (damondouglas) [inline playground/terraform/infrastructure/variables.tf:170]: Personally I found the `pg` in this `pg_location` name confusing initially.  May we consider variable names in the language of the cloud environment?  For example, `gke_location` is clear in that it refers to the location of the GKE node.  In some infrastructure-as-code modules, I've even seen just the use of `region` as it is conventionally understood that region refers to the Compute Engine region.  Then, when applying to the terraform GKE module, the location property is simply assigned to the `var.region` variable.

**Review by damondouglas — CHANGES_REQUESTED** [review]
@MakarkinSAkvelon (cc: @olehborysevych )

File: playground/terraform/README.md [hunk 22]
```diff
@@ -16,74 +16,179 @@
...
+https://cloud.google.com/iam/docs/granting-changing-revoking-access
 ```
+**Description:** * *It will explain how to assign required roles for your service account* *
 
-## 1. Create new environment
+6. How to install gcloud CLI to your local environment
+```
+https://cloud.google.com/sdk/docs/install?hl=en
+```
+**Description:** * *GCloud - Google command line interface, which will allow you to interact with the Google cloud platform thru the command line (https://en.wikipedia.org/wiki/Command-line_interface)* *
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:58]: Removed

File: playground/terraform/infrastructure/appengine/main.tf [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+resource "google_app_engine_application" "app_playground" {
+  project     = var.project_id
+  location_id = var.region == "us-central1" ? var.location_id_us : var.region
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/infrastructure/appengine/main.tf:18]: There is a problem with locations: for europe-west1 and us-central1 location name for AppEngine should be europe-west and us-central. So, this option (line 18) just helps us to identify and put correct location for AppEngine location:
https://cloud.google.com/appengine/docs/standard/locations#:~:text=If%20you%20already%20created%20an,the%20top%20of%20the%20page.

File: .github/workflows/playground_deploy_backend_gke.yml (hunk 16 above)
Comment (olehborysevych) [reply .github/workflows/playground_deploy_backend_gke.yml:73]: @MakarkinSAkvelon S3D_* keys are still here. Are we going to leave them for now?

File: playground/terraform/README.md [hunk 24]
```diff
@@ -16,74 +16,179 @@
...
+https://cloud.google.com/apis
+```
+**Description:** * *Short description of all Google Platform APIs* *
 
-* New environment folder
+8. How to Enable Google Cloud Platform API
+```
+https://cloud.google.com/endpoints/docs/openapi/enable-api
+```
+**Description:** * *Steps to enable APIs in Google Cloud Platform* *
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Fixed.
Add a new module which enable required APIs for BEAM Playground

File: playground/terraform/README.md [hunk 25]
```diff
@@ -16,74 +16,179 @@
...
 
-```bash
-echo 'bucket = "put your state bucket name here"' > /path/to/beam/playground/terraform/environment/{env-name}/state.tfbackend
-```
+These Google APIs will be needed for the BEAM Playground
+   - **Identity and Access Management (IAM)** - Manages identity and access control for Google Cloud Platform resources, including the creation of service accounts, which you can use to authenticate to Google and make API calls
+   - **Compute Engine API** - Creates and runs Virtual Machines on Google Cloud
+   - **App Engine Admin API** - Provisions and manages App Engine applications
+   - **Cloud Resource Manager API** - Creates, reads, and updates metadata for Google Cloud Platform resource containers
+You may find the instruction "How to enable Google Cloud Platform API" above
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Fixed.
Add a new module which enable required APIs for BEAM Playground

File: playground/terraform/infrastructure/variables.tf (hunk 21 above)
Comment (MakarkinSAkvelon) [reply playground/terraform/infrastructure/variables.tf:170]: Fixed

File: playground/terraform/infrastructure/appengine/main.tf [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    ]
+  count      = var.create_default_service ? 1 : 0
+  service    = "default"
+  version_id = "mlflow-default"
+  runtime    = "custom"
+  project    = var.project_id
+
+  deployment {
+    container {
+      image = "gcr.io/cloudrun/hello"
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/infrastructure/appengine/main.tf]: Fixed.

File: playground/terraform/README.md [hunk 27]
```diff
@@ -16,74 +16,179 @@
...
-# Requirements
+## Prerequisites:
 
-The following items need to be setup for the Playground cluster deployment on GCP:
+### Before you will start with the installation steps, please read this carefully:
+1. How to create a new project in Google Cloud Platform:
+```
+https://cloud.google.com/resource-manager/docs/creating-managing-projects
+```
+**Description:** * *How to create a new project in Google Cloud Platform. It's strongly recommended to use the new Google Cloud Project for your new projects* *
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Fixed

File: playground/terraform/README.md [hunk 28]
```diff
@@ -16,74 +16,179 @@
...
+https://cloud.google.com/iam/docs/creating-managing-service-account-keys
+```
+**Description:** * *In this instruction will be explained how to create a JSON key for your service account. it will be required for authentication.* *
 
-## 0. Create GCS bucket for state
+4. How to create a new bucket:
+```
+https://cloud.google.com/storage/docs/creating-buckets
+```
+**Description:** * *Short instruction for bucket creation. You will need at least one bucket for file keeping* *
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Yes, we need a bucket for tfstate file

File: playground/terraform/README.md [hunk 29]
```diff
@@ -16,74 +16,179 @@
...
-To deploy Playground infrastructure run gradle task:
+3. How to install HELM
+* [HELM link](https://helm.sh/docs/intro/install/) instructions for Windows and Linux operation systems
 
-```bash
-./gradlew playground:terraform:InitInfrastructure -Pproject_environment="env-name"
-```
+4. How to install GIT
+* [Git link](https://git-scm.com/download/win) instructions for Windows operation systems
+* [Git Link](https://git-scm.com/download/linux) instructions for Linux operation systems
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:72]: Fixed

File: playground/terraform/README.md [hunk 30]
```diff
@@ -16,74 +16,179 @@
...
+5. How to install Docker
+* [Docker link](https://docs.docker.com/engine/install/) instructions for Windows and Linux operation systems
 
-To deploy application run following steps:
+6. How to install GCloud
+* [Gcloud link](https://cloud.google.com/sdk/docs/install#linux) instructions for Windows and Linux operation systems
 
-* Authinticate in Artifact registry
+7. How to install Terraform
+* [Terraform link](https://www.terraform.io/downloads)
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Fixed

File: playground/terraform/README.md [hunk 31]
```diff
@@ -16,74 +16,179 @@
...
 
-```bash
-gcloud auth configure-docker us-central1-docker.pkg.dev
-```
+After installations of the required packages will be completed, you will need to download ***BEAM Playground*** from the repository. Please open the terminal and execute the following command:
 
-* Вeploy backend services
+`git clone 'url_to_github_project_with_beam_playground'`
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/README.md]: Fixed

File: playground/terraform/README.md [hunk 32]
```diff
@@ -16,74 +16,114 @@
...
+```
+gcloud container clusters get-credentials --region `chosen_pg_location` `gke_name` --project `project_id`
+```
+* We need to create database indexes for BEAM playground examples by the following command:
+```
+gcloud app deploy playground/index.yaml --project=`project_id`
+```
+That's all, the configuration of the environment has been completed. For deploying the backend part to the Google cloud Kubernetes engine, please execute the following command (Ensure you are in the "beam" folder):
+```
+./gradlew playground:terraform:gkebackend -Pproject_environment="env" -Pdocker-tag="tag" (env - folder name which you created for configuration files, tag - image tag for backend)
```
Comment (damondouglas) [inline playground/terraform/README.md:121]: Getting error:

```
Caused by: java.io.IOException: Cannot run program "helm" (in directory "<path to project>/playground/terraform"): error=2, No such file or directory
	at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:25)
	... 6 more
Caused by: java.io.IOException: error=2, No such file or directory
	... 7 more

```

When running, `helm version`, I see:

```
version.BuildInfo{Version:"v3.9.4", GitCommit:"dbc6d8e20fe1d58d50e6ed30f09a04a77e4c68db", GitTreeState:"clean", GoVersion:"go1.17.13"}
```

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: playground/terraform/README.md (hunk 32 above)
Comment (AKosolapov) [reply playground/terraform/README.md:121]: @damondouglas not clear from error message snippet - was command run from beam folder (per note above)?

Comment (MakarkinSAkvelon) [reply playground/terraform/README.md:121]: Hello, Damon
I hope you're switching branches from the master to the current one. Because, yes, till we merge it to the branch, the helm folder will not be there

File: playground/terraform/infrastructure/appengine/main.tf (hunk 26 above)
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply playground/terraform/infrastructure/appengine/main.tf:42]: Was forced to return back this option because of the following error during Frontend deployment:
Error: Error creating FlexibleAppVersion: googleapi: Error 400: The first service (module) you upload to a new application must be the 'default' service (module). Please upload a version of the 'default' service (module) before uploading a version for the 'frontend-stg' service (module). See the documentation for more information. Python: (https://developers.google.com/appengine/docs/python/modules/#Python_Uploading%%20modules) Java: (https://developers.google.com/appengine/docs/java/modules/#Java_Uploading%%20modules)

File: .github/workflows/playground_deploy_infrastructure.yml [hunk 33]
```diff
@@ -0,0 +1,18 @@
...
+  Infrastructure_deployment:
+    runs-on: ubuntu-latest
+    env:
+      PROJECT_ENV: beta
+    steps:
+      - name: git clone repo
+        uses: actions/checkout@v2
+      - name: Setup GCP account
+        run: |
+             echo "${{ secrets.S3D_GCP_PLAYGROUND_SA_KEY }}" | base64 -d > /tmp/gcp_access.json
```
[the remark this answers is not in the document]
Comment (MakarkinSAkvelon) [reply .github/workflows/playground_deploy_infrastructure.yml:14]: Fixed
